[Marketplace Catalog - Search] resource manager api

gregoks opened the pull request:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [x] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  4. If updating an existing version, please select the specific language SDKs and CLIs that must be refreshed after the swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No refresh required for updates in this PR

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist 

> **Applicability**: :warning: 
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs 
> - Adding new properties 
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [x] Check this box if any of the following appy to the PR so that the label "ARMReview" and "WaitForARMFeedback" will be added by bot to kick off ARM API Review.  Missing to check this box in the following scenario may result in delays to the ARM manifest review and deployment.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you are using OpenAPIHub to initialize the PR for adding a new version. More details, refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).

- [x] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If any of the following scenarios apply to the PR, request approval from the Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 
- [ ] Removing API(s) in a stable version
- [ ] Removing properties in a stable version
- [ ] Removing API version(s) in a stable version
- [ ] Updating API in a stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

Commit:
Added search data-plane

Commit:
added marketplace catalog resource manager

Commit:
fixed readme

Commit:
Fixed Spec Linter validations

Commit:
fixed all validation tools

Commit:
fixed avo

Commit:
fixed prettier check

Commit:
fixed spelling

Commit:
Removed search data-plane from resource-manager

anuchandy commented:
Hi @akning-ms, this is a resource manager pr.  Could you help redirect this to someone from the management-spec API review team

mentat9 commented on specification/marketplacecatalog/resource-manager/Search/stable/2022-02-02/search.json:
```diff
@@ -0,0 +1,785 @@
+              "$ref": "#/definitions/SearchResponseDto"
+            }
+          },
+          "400": {
```
>400 [](http://example.com/codeflow?start=11&length=3)

Don't explicitly define status code above 2XX. Put all of your error logic in the default branch.

gregoks replied:
@mentat9  Not sure how can I put both 400 and 500 under default section. Can you please share an example?

mentat9 commented on specification/marketplacecatalog/resource-manager/Search/stable/2022-02-02/search.json:
```diff
@@ -0,0 +1,785 @@
+              "$ref": "#/definitions/SearchErrorResponseDto"
+            }
+          },
+          "500": {
```
>500 [](http://example.com/codeflow?start=11&length=3)

Use the default branch.

gregoks replied:
Done

mentat9 commented on specification/marketplacecatalog/resource-manager/Search/stable/2022-02-02/search.json:
```diff
@@ -0,0 +1,785 @@
+          "200": {
+            "description": "OK. The request has succeeded.",
+            "schema": {
+              "$ref": "#/definitions/OperationListResult"
```
>#/definitions/OperationListResult [](http://example.com/codeflow?start=23&length=33)

Use types from v3 common types unless you have a compelling reason not to.

gregoks replied:
@mentat9 can you please share an example of such usage?

mentat9 commented on specification/marketplacecatalog/resource-manager/Search/stable/2022-02-02/search.json:
```diff
@@ -0,0 +1,785 @@
+          "default": {
+            "description": "In case internal error",
+            "schema": {
+              "$ref": "#/definitions/SearchErrorResponseDto"
```
>#/definitions/SearchErrorResponseDto [](http://example.com/codeflow?start=23&length=36)

Please use ErrorResponse from v3 common types unless you have a compelling reason not to. The extra properties in the common definition are all optional, you can leave them out of the response if you prefer. If you want to use them later, they will be available.

gregoks replied:
@mentat9 can you please share an example how to use v3 common types and not v2?

gregoks replied:
done

mentat9 commented on specification/marketplacecatalog/resource-manager/Search/stable/2022-02-02/search.json:
```diff
@@ -0,0 +1,785 @@
+          },
+          {
+            "type": "string",
+            "name": "api-version",
```
>api-version [](http://example.com/codeflow?start=21&length=11)

Use api-version parameter from v3 common types for this.

gregoks replied:
done

mentat9 commented on specification/marketplacecatalog/resource-manager/Search/stable/2022-02-02/search.json:
```diff
@@ -0,0 +1,785 @@
+        "operationId": "Operations_List",
+        "parameters": [
+          {
+            "$ref": "#/parameters/ApiVersionParameter"
```
>#/parameters/ApiVersionParameter [](http://example.com/codeflow?start=21&length=32)

Use api-version from v3 common types

gregoks replied:
done

mentat9 commented on specification/marketplacecatalog/resource-manager/Search/stable/2022-02-02/search.json:
```diff
@@ -0,0 +1,785 @@
+    "OperationProperties": {
+      "description": "Operation properties.",
+      "type": "object",
+      "properties": {}
```
>properties": {} [](http://example.com/codeflow?start=7&length=15)

ARM recommends against empty property bags or unschematized objects. This won't be there if you switch to v3 common types.

gregoks replied:
done

mentat9 commented on specification/marketplacecatalog/resource-manager/Search/stable/2022-02-02/search.json:
```diff
@@ -0,0 +1,785 @@
+  ],
+  "paths": {
+    "/providers/Microsoft.Marketplace/search": {
+      "get": {
```
>get [](http://example.com/codeflow?start=7&length=3)

RPC requires that tenant-level RO APIs are implemented using POST instead of GET. Ref: [https://github.com/Azure/azure-resource-manager-rpc/blob/a7dd5c0021d0a84bcad6d4736e3fcc3b319b20af/v1.0/proxy-api-reference.md#request-tenant-wide-operations](https://github.com/Azure/azure-resource-manager-rpc/blob/a7dd5c0021d0a84bcad6d4736e3fcc3b319b20af/v1.0/proxy-api-reference.md#request-tenant-wide-operations).

gregoks replied:
@mentat9 that's not tenant based API. we have several similar API's that come from marketplace that are "GET"

mentat9 replied:
Tenant-level, in the sense that it applies across all subscriptions in a tenant, rather than applying to a single subscription (or RG). That is exactly the case covered in the RPC (linked above).

The current Microsoft.Marketplace swagger doesn't document other tenant-level RO APIs using GET (other than operations, which is a special ARM api). The only tenant-level GET documented for your RP is a full resource type supporting GET/PUT/PATCH/DELETE. The search API is a read-only operation which should be POST per the RPC.

gregoks replied:
got it. Changed to POST.

mentat9 commented on specification/marketplacecatalog/resource-manager/Search/stable/2022-02-02/search.json:
```diff
@@ -0,0 +1,785 @@
+        "displayName": {
+          "type": "string"
+        },
+        "isAzureBenefitEligible": {
```
>isAzureBenefitEligible [](http://example.com/codeflow?start=9&length=22)

There are a bunch of booleans in this response. ARM recommends string enums over boolean. Generally enums make better properties than booleans. They are more descriptive, flexible and future-proof while being easier for customers to discover, understand and use. String enums allow adding values without a new API version, which looks like it could be a potential benefit for some of these.

Ref: [https://armwiki.azurewebsites.net/rp_onboarding/process/api_review_best_practices.html?q=boolean#common-issues-found-in-review](https://armwiki.azurewebsites.net/rp_onboarding/process/api_review_best_practices.html?q=boolean#common-issues-found-in-review).

As an enum, this property could be something like:

benefitEligibility: { Default (or None), Azure }.

gregoks replied:
@mentat9 I understand your point but unfortunately this property is based on cosmosDB property that has similar name so I would want to keep the same naming and type for this field so it wont confuse our clients that use both Search API and other API's that get data from our cosmos.

mentat9 commented on specification/marketplacecatalog/resource-manager/Search/stable/2022-02-02/search.json:
```diff
@@ -0,0 +1,785 @@
+        "isAzureBenefitEligible": {
+          "type": "boolean"
+        },
+        "isCoreVm": {
```
>isCoreVm [](http://example.com/codeflow?start=9&length=8)

Could be vmType: { Default (or Auto?), Core } or similar.

gregoks replied:
@mentat9 I understand your point but unfortunately this property is based on cosmosDB property that has similar name so I would want to keep the same naming and type for this field so it wont confuse our clients that use both Search API and other API's that get data from our cosmos.

mentat9 commented on specification/marketplacecatalog/resource-manager/Search/stable/2022-02-02/search.json:
```diff
@@ -0,0 +1,785 @@
+        "isCoreVm": {
+          "type": "boolean"
+        },
+        "isPreferredSolution": {
```
>isPreferredSolution [](http://example.com/codeflow?start=9&length=19)

Could be solutionType: { Standard, Preferred } or similar.

gregoks replied:
@mentat9 I understand your point but unfortunately this property is based on cosmosDB property that has similar name so I would want to keep the same naming and type for this field so it wont confuse our clients that use both Search API and other API's that get data from our cosmos.

mentat9 commented on specification/marketplacecatalog/resource-manager/Search/stable/2022-02-02/search.json:
```diff
@@ -0,0 +1,785 @@
+        "isPreferredSolution": {
+          "type": "boolean"
+        },
+        "isAdditionalPurchaseRequired": {
```
>isAdditionalPurchaseRequired [](http://example.com/codeflow?start=9&length=28)

Could be requirements: { Default/None, AdditionalPurchase } or similar.

gregoks replied:
@mentat9 I understand your point but unfortunately this property is based on cosmosDB property that has similar name so I would want to keep the same naming and type for this field so it wont confuse our clients that use both Search API and other API's that get data from our cosmos.

mentat9 commented on specification/marketplacecatalog/resource-manager/Search/stable/2022-02-02/search.json:
```diff
@@ -0,0 +1,785 @@
+        "isAdditionalPurchaseRequired": {
+          "type": "boolean"
+        },
+        "isPowerBICertified": {
```
>isPowerBICertified [](http://example.com/codeflow?start=9&length=18)

Could be certification: { None, PowerBI } or similar.

gregoks replied:
@mentat9 I understand your point but unfortunately this property is based on cosmosDB property that has similar name so I would want to keep the same naming and type for this field so it wont confuse our clients that use both Search API and other API's that get data from our cosmos.

mentat9 commented on specification/marketplacecatalog/resource-manager/Search/stable/2022-02-02/search.json:
```diff
@@ -0,0 +1,785 @@
+        "isPowerBICertified": {
+          "type": "boolean"
+        },
+        "isIndustryCloud": {
```
>isIndustryCloud [](http://example.com/codeflow?start=9&length=15)

Could be cloudType: { Default, Industry } or similar.

gregoks replied:
@mentat9 I understand your point but unfortunately this property is based on cosmosDB property that has similar name so I would want to keep the same naming and type for this field so it wont confuse our clients that use both Search API and other API's that get data from our cosmos.

mentat9 commented on specification/marketplacecatalog/resource-manager/Search/stable/2022-02-02/search.json:
```diff
@@ -0,0 +1,785 @@
+        "isIndustryCloud": {
+          "type": "boolean"
+        },
+        "isMicrosoftProduct": {
```
>isMicrosoftProduct [](http://example.com/codeflow?start=9&length=18)

Could be productType: { Microsoft, ThirdParty (or External) } or similar.

gregoks replied:
@mentat9 I understand your point but unfortunately this property is based on cosmosDB property that has similar name so I would want to keep the same naming and type for this field so it wont confuse our clients that use both Search API and other API's that get data from our cosmos.

mentat9 commented on specification/marketplacecatalog/resource-manager/Search/stable/2022-02-02/search.json:
```diff
@@ -0,0 +1,785 @@
+        "isMicrosoftProduct": {
+          "type": "boolean"
+        },
+        "isPreview": {
```
>isPreview [](http://example.com/codeflow?start=9&length=9)

Could be readiness: { Preview, Release }, or similar.

gregoks replied:
@mentat9 I understand your point but unfortunately this property is based on cosmosDB property that has similar name so I would want to keep the same naming and type for this field so it wont confuse our clients that use both Search API and other API's that get data from our cosmos.

mentat9 commented on specification/marketplacecatalog/resource-manager/Search/stable/2022-02-02/search.json:
```diff
@@ -0,0 +1,785 @@
+    "PlanDto": {
+      "type": "object",
+      "properties": {
+        "cspState": {
```
>cspState [](http://example.com/codeflow?start=9&length=8)

Consider expanding the acronym in the name.

gregoks replied:
@mentat9 I understand your point but unfortunately this property is based on cosmosDB property that has similar name so I would want to keep the same naming and type for this field so it wont confuse our clients that use both Search API and other API's that get data from our cosmos.

gregoks commented on specification/marketplacecatalog/resource-manager/Search/stable/2022-02-02/search.json:
```diff
@@ -0,0 +1,785 @@
+    }
+  },
+  "parameters": {
+    "ApiVersionParameter": {
```
@mentat9  can you please show an example? If you are saying that my json can be open api v3 then one of the validators failed such scheme because it demanded it to be v2..

gregoks replied:
ok I think I figured it out. done

Commit:
fixed code review comments

akning-ms commented:
> Hi @akning-ms, this is a resource manager pr. Could you help redirect this to someone from the management-spec API review team

Yes. removed you from assignee,   @gregoks next time, the change to readme in dataplane is by mistaken? next time better to send the PR which only for management plane or dataplane, not mix them in one PR to avoid confusion, dataplane & management PR has different review process.

Commit:
fixed code review

Commit:
removed 400 error

gregoks commented:
> > Hi @akning-ms, this is a resource manager pr. Could you help redirect this to someone from the management-spec API review team
> 
> Yes. removed you from assignee, @gregoks next time, the change to readme in dataplane is by mistaken? next time better to send the PR which only for management plane or dataplane, not mix them in one PR to avoid confusion, dataplane & management PR has different review process.

Yea the dataplane readme was mistake, I removed it from this PR.

Commit:
removed error codes from example

Commit:
added default api version value

Commit:
fixed prettier and removed default api version

Commit:
changed ARM from GET to POST

Commit:
changed private store params to body

Commit:
fixed body param

Commit:
fixed post example

mentat9 commented:
@gregoks - Signed off from ARM.
